library/pkcs7.c

```diff
@@ -184,13 +187,13 @@ static int pkcs7_get_certificates(unsigned char **p, unsigned char *end,
     size_t len2 = 0;
     unsigned char *end_set, *end_cert, *start;
 
-    if ((ret = mbedtls_asn1_get_tag(p, end, &len1, MBEDTLS_ASN1_CONSTRUCTED
-                                    | MBEDTLS_ASN1_CONTEXT_SPECIFIC)) != 0) {
-        if (ret == MBEDTLS_ERR_ASN1_UNEXPECTED_TAG) {
-            return 0;
-        } else {
-            return MBEDTLS_ERROR_ADD(MBEDTLS_ERR_PKCS7_INVALID_FORMAT, ret);
-        }
+    ret = mbedtls_asn1_get_tag(p, end, &len1, MBEDTLS_ASN1_CONSTRUCTED
+                               | MBEDTLS_ASN1_CONTEXT_SPECIFIC);
+    if (ret == MBEDTLS_ERR_ASN1_UNEXPECTED_TAG) {
+        return 0;
```

Comment:
We've got an unexpected tag, but we return 0 ("0 or higher is valid") - what am I missing?

Reply:
The entire ExtendedCertificatesAndCertificates field is optional, and could be omitted if there are zero certificates present. In that case we hit this code path. See https://www.rfc-editor.org/rfc/rfc2315#:~:text=contentInfo%20ContentInfo%2C%0A%20%20%20%20%20certificates-,%5B0%5D%20IMPLICIT%20ExtendedCertificatesAndCertificates%0A%20%20%20%20%20%20%20%20%20%20OPTIONAL%2C,-crls%0A%20%20%20%20%20%20%20%5B1%5D%20IMPLICIT